File: java/linkedlist/DeleteMiddleNode.java
/**
 * Q14. Delete Middle Node of Linked List
 * 
 * Problem:
 * You are given the head of a linked list. Delete the middle node, and return the head of the modified linked list.
 * The middle node of a linked list of size n is the ⌊n / 2⌋th node from the start using 0-based indexing,
 * where ⌊x⌋ denotes the largest integer less than or equal to x.
 * For n = 1, 2, 3, 4, and 5, the middle nodes are 0, 1, 1, 2, and 2, respectively.
 * 
 * Example:
 * Input: head = [1,3,4,7,1,2,6]
 * Output: [1,3,4,1,2,6]
 * 
 * Approach:
 * 1. Use slow and fast pointers to find the middle node
 * 2. Keep track of the previous node of slow pointer
 * 3. Delete the middle node by connecting prev.next to slow.next
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class DeleteMiddleNode {
    public ListNode deleteMiddle(ListNode head) {
        // Handle edge cases
        if (head == null || head.next == null) {
            return null;
        }

        // Initialize pointers
        ListNode prev = null;
        ListNode fast = head;
        ListNode slow = head;

        // Find middle node
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        // Delete middle node
        prev.next = slow.next != null ? slow.next : null;
        
        return head;
    }

    // Helper method to print the list
    private static void printList(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    // Main method for testing
    public static void main(String[] args) {
        // Test case 1: List with odd number of nodes [1,3,4,7,1,2,6]
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(3);
        head1.next.next = new ListNode(4);
        head1.next.next.next = new ListNode(7);
        head1.next.next.next.next = new ListNode(1);
        head1.next.next.next.next.next = new ListNode(2);
        head1.next.next.next.next.next.next = new ListNode(6);

        DeleteMiddleNode solution = new DeleteMiddleNode();
        System.out.println("Original list:");
        printList(head1);
        
        ListNode result1 = solution.deleteMiddle(head1);
        System.out.println("List after deleting middle node:");
        printList(result1);

        // Test case 2: List with even number of nodes [1,2,3,4]
        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(2);
        head2.next.next = new ListNode(3);
        head2.next.next.next = new ListNode(4);

        System.out.println("\nOriginal list:");
        printList(head2);
        
        ListNode result2 = solution.deleteMiddle(head2);
        System.out.println("List after deleting middle node:");
        printList(result2);

        // Test case 3: List with two nodes [1,2]
        ListNode head3 = new ListNode(1);
        head3.next = new ListNode(2);

        System.out.println("\nOriginal list:");
        printList(head3);
        
        ListNode result3 = solution.deleteMiddle(head3);
        System.out.println("List after deleting middle node:");
        printList(result3);

        // Test case 4: Single node [1]
        ListNode head4 = new ListNode(1);

        System.out.println("\nOriginal list:");
        printList(head4);
        
        ListNode result4 = solution.deleteMiddle(head4);
        System.out.println("List after deleting middle node:");
        printList(result4);

        // Test case 5: Empty list
        ListNode result5 = solution.deleteMiddle(null);
        System.out.println("\nEmpty list result: " + (result5 == null ? "null" : "not null"));
    }
} 

File: javascript/linkedlist/DeleteMiddleNode.js
/**
 * Delete Middle Node
 * 
 * Problem:
 * Given the head of a linked list, delete the middle node, and return the head of the modified linked list.
 * The middle node of a linked list of size n is the ⌊n / 2⌋th node from the start using 0-based indexing.
 * For n = 1, 2, 3, 4, and 5, the middle nodes are 0, 1, 1, 2, and 2, respectively.
 * 
 * Example:
 * Input: head = [1,3,4,7,1,2,6]
 * Output: [1,3,4,1,2,6]
 * Explanation:
 * - The middle node is 7 (index 3)
 * - After deleting it, the list becomes [1,3,4,1,2,6]
 * 
 * Approach:
 * 1. Use slow and fast pointers to find the middle node
 * 2. Keep track of the previous node of the middle node
 * 3. Delete the middle node by updating pointers
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class DeleteMiddleNode {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    deleteMiddle(head) {
        if (!head || !head.next) return null;

        let slow = head;
        let fast = head;
        let prev = null;

        while (fast && fast.next) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        // Delete the middle node
        prev.next = slow.next;

        return head;
    }

    // Helper method to create a linked list
    static createList(values) {
        if (!values || values.length === 0) return null;
        
        const head = new ListNode(values[0]);
        let current = head;
        
        for (let i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }
        
        return head;
    }

    // Helper method to print the list
    static printList(head) {
        let current = head;
        const values = [];
        while (current) {
            values.push(current.val);
            current = current.next;
        }
        console.log(values.join(' -> '));
    }
}

// Test cases
function runTests() {
    const solution = new DeleteMiddleNode();

    // Test case 1: Regular case
    console.log("Test case 1: Regular case");
    const list1 = DeleteMiddleNode.createList([1, 3, 4, 7, 1, 2, 6]);
    console.log("Original list:");
    DeleteMiddleNode.printList(list1);
    console.log("List after deleting middle node:");
    DeleteMiddleNode.printList(solution.deleteMiddle(list1));

    // Test case 2: Two node list
    console.log("\nTest case 2: Two node list");
    const list2 = DeleteMiddleNode.createList([1, 2]);
    console.log("Original list:");
    DeleteMiddleNode.printList(list2);
    console.log("List after deleting middle node:");
    DeleteMiddleNode.printList(solution.deleteMiddle(list2));

    // Test case 3: Three node list
    console.log("\nTest case 3: Three node list");
    const list3 = DeleteMiddleNode.createList([1, 2, 3]);
    console.log("Original list:");
    DeleteMiddleNode.printList(list3);
    console.log("List after deleting middle node:");
    DeleteMiddleNode.printList(solution.deleteMiddle(list3));

    // Test case 4: Single node list
    console.log("\nTest case 4: Single node list");
    const list4 = DeleteMiddleNode.createList([1]);
    console.log("Original list:");
    DeleteMiddleNode.printList(list4);
    console.log("List after deleting middle node:");
    DeleteMiddleNode.printList(solution.deleteMiddle(list4));

    // Test case 5: Empty list
    console.log("\nTest case 5: Empty list");
    const list5 = null;
    console.log("Original list: null");
    console.log("List after deleting middle node:");
    DeleteMiddleNode.printList(solution.deleteMiddle(list5));

    // Test case 6: List with even number of nodes
    console.log("\nTest case 6: List with even number of nodes");
    const list6 = DeleteMiddleNode.createList([1, 2, 3, 4]);
    console.log("Original list:");
    DeleteMiddleNode.printList(list6);
    console.log("List after deleting middle node:");
    DeleteMiddleNode.printList(solution.deleteMiddle(list6));

    // Test case 7: List with negative numbers
    console.log("\nTest case 7: List with negative numbers");
    const list7 = DeleteMiddleNode.createList([-1, -2, -3, -4, -5]);
    console.log("Original list:");
    DeleteMiddleNode.printList(list7);
    console.log("List after deleting middle node:");
    DeleteMiddleNode.printList(solution.deleteMiddle(list7));
}

// Run the tests
runTests(); 